generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,157 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by bitwise shifting left or right by an
+   * arbitrary number of bits. Generates an opaque value if the type of zero to be generated is an
+   * integer and the shading language in use supports bitwise operations.
+   *
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // We need to make sure the LHS and RHS of the shift expression are based on the same type
+    // explicitly, so we can't generate opaque values to shift by as they will implicitly be
+    // considered signed integers, and both sides of the shift must be the same type.
+    final Expr shiftValue;
+    if (isZero) {
+      // We can't shift zero reliably if the number is signed because of sign bit extension, so
+      // we simply shift it zero if it's signed.
+      shiftValue = type.getElementType() == BasicType.INT
+          ? new IntConstantExpr("0")
+          : new UIntConstantExpr(generator.nextInt(1000) + "u");
```

Comment:
Now, rather than generating a literal in the range [0, 9], it would be more adventurous to generate a random expression and clamp it into the range [0, 9).

I would do something like:

identity(clamp(fuzzed_expr, opaque_zero, identity(9))

for the RHS.

Reply:
A problem with clamping into the range [0, 9) is that the result will almost certainly be 8.  An alternative that might be better is to pick a value n, with 0 <= n <= 9, and clamp into the range [0, n).  The result will almost certainly be (n-1), but since n is chosen randomly this creates more variation.

Reply:
Done.